crates/apollo-compiler/src/validation/mod.rs

```diff
@@ -266,6 +287,40 @@ impl DiagnosticData {
     }
 }
 
+impl<'a> Diagnostic<'a> {
+    /// Get the line and column number where this diagnostic was raised.
+    pub fn get_line_column(&self) -> Option<GraphQLLocation> {
+        // let source = self.sources.get(&id)?;
+        // let char_index = source.map_index(index);
+        // let (_, line, column) = source.ariadne.get_offset_line(char_index)?;
+        // Some((line, column))
+        let loc = self.data.location?;
+        let source = self.sources.get(&loc.file_id)?;
+        source
+            .get_line_column(loc.offset())
+            .map(|(line, column)| GraphQLLocation {
+                line: line + 1,
+                column: column + 1,
+            })
+    }
+
+    /// Get serde_json serialisable version of the current diagnostic.
+    pub fn to_json(&self) -> GraphQLError {
+        let mut locations = vec![];
+
+        if let Some(location) = self.get_line_column() {
+            locations.push(location);
+        }
+
+        GraphQLError {
+            // TODO(@lrlna): this needs DiagnosticData to have a Display impl
+            // using only diagnostic messages.
+            message: self.to_string(),
```

Comment:
```suggestion
            message: self.data.details.to_string(),
```

This relies on the `Display` impl generated by `derive(thiserror::Error)`  based on attributes like `#[error("type extension for undefined type `{name}`")]`.

It the same that we pass to `ariadne::ReportBuilder::with_message`. Later we can match on each `Detail` variant and add more details that Ariadne reports have in labels/help/note, but this is a good starting point.

Reply:
The Router also ended up needing this:

```rust
    pub fn message(&self) -> &impl fmt::Display {
        &self.data.details
    }
```